### ISSUE TITLE
Make it possible to disable the use of flexible layout templates

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#4161]
Make it possible to disable the use of flexible layout templates (~should this be it's own submodule?~)

Phases:
* first: add a permission for flexible layout templates (urgent bug fix?)
* don't forget the update hook
* move layout builder to a separate module (less urgent UX improvement)

---
~~PR: https://github.com/backdrop/backdrop/pull/2952~~
PR by @docwilmot: https://github.com/backdrop/backdrop/pull/3108 (merged)

Follow-up by @klonos: https://github.com/backdrop/backdrop/pull/3140 (adds update hook to add permission for existing sites)

**Comment by stpaultim** [comment]
I had to think about the use case for flexible layout template permission. At first, I was thinking about it would be different from the normal layout permission, but the answer came quickly and clearly. 

Permission to "administer layouts" only allows the user to access enabled layouts. The site admin may create a define list of layouts for site editors to use. Flexible layouts would be a way to bypass that limitation. Without permissions for flexible layouts, it's not possible to limit editors to a select list of layouts. 

I agree, this is probably a pretty important feature. 

It's less clear to me why we would need to disable flexible layouts entirely once this permission is in place.

**Comment by jenlampton** [comment]
> The site admin may create a defined list of layouts for site editors to use. 

I do this often when only some of the my layouts have been tested with my theme. It's also useful when I'm the only person on the site -- to keep my future self out of trouble :)

> It's less clear to me why we would need to disable flexible layouts entirely once this permission is in place.

It's not necessary, but it would be an improvement. We can also "disable" the admin bar by using the "Access administration bar" permission. But I have a feeling that people who don't want it used on their site prefer to disable the module instead.

**Comment by jenlampton** [comment]
Is this a bug?

**Comment by jenlampton** [comment]
PR available for review: https://github.com/backdrop/backdrop/pull/2952

**Comment by jenlampton** [comment]
Tests are failing, bumping milestone

**Comment by docwilmot** [comment]
Updated this, tests green, awaiting code review.

**Comment by stpaultim** [comment]
Just to be clear, the current PR only addresses the issue with permissions correct? 

I tested that part and it seems to work as expected. It will be great to get this finished.

I assume that "move layout builder to a separate module" is a follow-up issue?

**Comment by docwilmot** [comment]
Yes I think this should be separate. Can I get a world for me?

**Comment by jenlampton** [comment]
Code changes look great, RTBC!

**Comment by quicksketch** [comment]
I merged https://github.com/backdrop/backdrop/pull/3108 into 1.x for 1.16.0. However, afterwards I realized we need an update hook here. If a user previously had "administer layouts" they should be given "administer flexible layouts" because they had that permission implicitly before it was separated.

Could we get a follow-up MR to add an update hook to layout.install?

**Comment by docwilmot** [comment]
Mmm good catch. Maybe "administer layouts" should be a superset of "administer flexible layouts". Wonder what happens if you grant "administer flexible layouts" but not "administer layouts". Cant test now but we need to look into that when fixing this.

**Comment by klonos** [comment]
PR to add the update hook for now, so that user roles in existing sites do not have permission regression: https://github.com/backdrop/backdrop/pull/3136

I'll look into what @docwilmot said next.

**Comment by klonos** [comment]
...OK, so granting "administer flexible templates" does not allow any access to `admin/structure/layouts` nor to `admin/structure/layouts/settings`, but the user can access `admin/structure/layouts/settings/flexible-template/add`. So they can create and configure new templates. They can also edit existing templates, if they know their machine name and manually navigate to the `admin/structure/layouts/settings/flexible-template/TEMPLATENAME/configure` path.

So, it seems that we want the "administer flexible templates" permission to behave the same as the "access administration bar" permission, which shows the following message:

>Also allow Authenticated role the "Use the administration pages" permission? This is usually required to display any pages within the administration bar.

**Comment by klonos** [comment]
I had a look, and it seems that the 2 "Use the administration pages" and "Access administration bar" permissions are linked to each other: selecting either one throws a message about selecting the other.

On my local, I have implemented a solution that does the same with the "Administer Layouts" and "Administer Flexible Templates" permissions, but I am thinking that perhaps in this case it shouldn't be a two-way thing. We should only be suggesting to enable "Administer Layouts" when "Administer Flexible Templates" is checked - not the other way.

Anyway, it sounds like this may need to be addressed in a follow-up.

**Comment by quicksketch** [comment]
@klonos Hm, tests are failing after that update hook. It seems that adding an update to any new module that did not exist in D7 results in update.php requiring to be run multiple times when upgrading from D7 to Backdrop.

I filed a modified version of your PR at https://github.com/backdrop/backdrop/pull/3139 that fixes this problem, any new modules enabled in the D7 -> Backdrop upgrade are assumed to be running at the latest schema version, instead of 0, which makes it so that update.php doesn't think it needs to be run again.

**Comment by jenlampton** [comment]
Nice solution @quicksketch!  I've created a follow-up issue so we can address the missing update hooks for backdrop-only modules more holistically, later: https://github.com/backdrop/backdrop-issues/issues/4396

**Comment by quicksketch** [comment]
Thanks @klonos! Merged https://github.com/backdrop/backdrop/pull/3140 into 1.x, which put the update hook in system.install. We have a separate issue at https://github.com/backdrop/backdrop-issues/issues/1759 to solve the problem of Layout module (and any other core module that did not exist in D7) not being able to contain update hooks in their own .install files.

**Comment by klonos** [comment]
Follow-up task to split this to a separate module: #4402
